New: max-statements-per-line (fixes #5424)

scriptdaemon opened the pull request:
Added new rule (`max-statements-per-line`) to enforce a maximum number of statements per line.

michaelficarra commented on lib/rules/max-statements-per-line.js:
```diff
@@ -0,0 +1,119 @@
+
+    return {
+        "Program": enforceMaxStatementsPerLine,
+        "BlockStatement": enforceMaxStatementsPerLine
```
You've missed SwitchCase and possibly others.

michaelficarra commented on lib/rules/max-statements-per-line.js:
```diff
@@ -0,0 +1,119 @@
+     * @private
+     */
+    function isStatementsOnSameLine(node1, node2) {
+        if (!node1 || !node2) {
```
You should never pass a falsy value to this function.

michaelficarra commented on lib/rules/max-statements-per-line.js:
```diff
@@ -0,0 +1,119 @@
+     * @returns {void}
+     * @private
+     */
+    function report(node, count, max) {
```
This function doesn't need to take max as a parameter.

michaelficarra commented on lib/rules/max-statements-per-line.js:
```diff
@@ -0,0 +1,119 @@
+                numberOfStatementsOnThisLine = 1;
+                lastStatementLine = currentStatement.loc.end.line;
+            }
+            if (!isStatementsOnSameLine(currentStatement, nextStatement) &&
```
This call isn't necessary.

michaelficarra commented on lib/rules/max-statements-per-line.js:
```diff
@@ -0,0 +1,119 @@
+                lastStatementLine = currentStatement.loc.end.line;
+            }
+            if (!isStatementsOnSameLine(currentStatement, nextStatement) &&
+                    numberOfStatementsOnThisLine > maxStatementsPerLine) {
```
With this approach, you're going to make a report for each number over the max. Users probably only want to have one report per line, so you should use an equality comparison.

scriptdaemon replied:
That's what `!isStatementsOnSameLine` is for.

scriptdaemon replied:
That's what `!isStatementsOnSameLine` is for. It verifies that the node being reported is the last on the line (so the error reports the number of statements on the line, not just the count one over the max).

michaelficarra replied:
I'm not sure the user cares how many statements are on the line; just that the rule has been violated.

michaelficarra commented on lib/rules/max-statements-per-line.js:
```diff
@@ -0,0 +1,119 @@
+        var lastStatementLine = body[0].loc.end.line;
+        for (var i = 1, l = body.length; i < l; ++i) {
+            var currentStatement = body[i];
+            var nextStatement = body[i + 1];
```
You're going to make an out of bounds array index here.

michaelficarra commented:
There's no tests with a max of 0, and there's no examples in the docs with a max of anything other than 1.

nzakas commented on docs/rules/max-statements-per-line.md:
```diff
@@ -0,0 +1,67 @@
+# Limit Maximum Number of Statements Per Line (max-statements-per-line)
+
+The `max-statements-per-line` rule allows you to specify the maximum number of statements allowed per line.
```
This section is an intro to the concept that the rule tests, not about the rule itself. See http://eslint.org/docs/rules/max-nested-callbacks as an example

nzakas commented on docs/rules/max-statements-per-line.md:
```diff
@@ -0,0 +1,67 @@
+
+This rule allows you to configure the maximum number of statements allowed per line.  The default is 1.
+
+## Options
```
@ilyavolodin should this be heading level 3 now?

scriptdaemon replied:
All other rules I used as examples do not make this a level 3 header.

ilyavolodin replied:
Nope, `options` still stays as level 2, however, each individual option should be level 3 heading now.

nzakas commented on docs/rules/max-statements-per-line.md:
````diff
@@ -0,0 +1,67 @@
+
+## Options
+
+```json
````
This isn't valid JSON, so will cause incorrect syntax highlighting. Just omit the "json"

scriptdaemon commented:
I believe I have addressed all concerns. I also updated the code to catch things like

```js
/*eslint max-statements-per-line: [2, 1]*/

if (condition) { statement; } else { statement; }
```

nzakas commented:
@michaelficarra have all of your concerns been addressed?

michaelficarra commented on docs/rules/max-statements-per-line.md:
```diff
@@ -0,0 +1,152 @@
+var bar = 1; var baz = 2;
+if (condition) var bar = 1; if (condition) var baz = 2;
+if (condition) { } if (condition) { }
+if (condition) { var bar = 1; } else { var bar = 1; }
```
How many statements are on this line? Why?

scriptdaemon replied:
Three.

1. The `IfStatement` (with or without an alternate)
2. The `VariableDeclaration` inside the `BlockStatement` of the consequent
3. The `VariableDeclaration` inside the `BlockStatement` of the alternate

`BlockStatement`s that comprise the body of another statement (`IfStatement` consequent or alternate, `ForStatement`, etc.) do not count as their own statements (otherwise an `if / else` with both `BlockStatement`s would count as three alone). `BlockStatement`s that stand on their own (e.g. parent is a `BlockStatement` or `Program`) will count as a statement, however.

Do you agree with this behavior?

scriptdaemon commented:
I simplified some of the code a bit and rebased with master.

michaelficarra commented:
I would like to see tests for these programs:

```js
(x => { ; }) ? (y => { ; }) : (z => { ; });
```

```js
function f({[x => { ; }]: y = z => { ; }}) {}
```

```js
switch(x) { case y: ; case z: ; }
```

```js
let x = x => { ; }, y = y => { ; };
```

```js
[x => { ; }, y => { ; }, z => { ; }];
```

```js
({ x: x => { ; }, y: y => { ; }, z : z => { ; }});
```

```js
f(x => { ; }, y => { ; }, z => { ; });
```

```js
x => { ; }, y => { ; }, z => { ; };
```

scriptdaemon commented:
@michaelficarra Done.

nzakas commented:
Just waiting for @michaelficarra to confirm all his concerns were addressed.

nzakas commented on docs/rules/README.md:
```diff
@@ -172,6 +172,7 @@ These rules are purely matters of style and are quite subjective.
 * [max-nested-callbacks](max-nested-callbacks.md) - specify the maximum depth callbacks can be nested
 * [max-params](max-params.md) - limits the number of parameters that can be used in the function declaration
 * [max-statements](max-statements.md) - specify the maximum number of statement allowed in a function
+* [max-statements-per-line](max-statements-per-line.md) - specify the maximum number of statements allowed per line
```
The description here should be the same as the title of rule doc page

nzakas commented on docs/rules/max-statements-per-line.md:
````diff
@@ -0,0 +1,152 @@
+Examples of **incorrect** code for a maximum of 1:
+
+```js
+/*eslint max-statements-per-line: [2, 1]*/
````
We don't do rules in this format anymore because it's unckear what 1 means. I'd like to see the rule be configured like this:

```
max-statements-per-line: ["error ", {
    max: 1
}]
```

That's more the way we've been doing rules lately.

scriptdaemon commented:
@nzakas All changes you requested have been added.

nzakas commented:
@pedrottimark would you mind looking over the docs in this PR? I'm not sure how we are doing options now, but it seems like this would need some sort of reorganizing.

pedrottimark commented:
@scriptdaemon Nice work!

Here are suggestions to make this the first new rule doc to start out leveled up with the rising tide:
* Because the rule has options, move example code from under h2 Rule Details to h3 max.
* Pair up the examples of incorrect/correct code to minimize scrolling when comparing.
* Just yesterday we changed a phrase in the example sentences to “for this rule with…”
* Delete the rule configuration JSON and move it into the sentence. Notice that the eslint comments in the examples also have a space within the braces.

**Options**

**max**

The `"max"` object property is optional (default: 1).

Examples of **incorrect** code for this rule with the default `{ "max": 1 }` option:
`/*eslint max-statements-per-line: ["error", { "max": 1 }]*/`

Examples of **correct** code for this rule with the default `{ "max": 1 }` option:
`/*eslint max-statements-per-line: ["error", { "max": 1 }]*/`

Examples of **incorrect** code for this rule with a sample `{ "max": 2 }` option:
`/*eslint max-statements-per-line: ["error", { "max": 2 }]*/`

Examples of **correct** code for this rule with a sample `{ "max": 2 }` option:
`/*eslint max-statements-per-line: ["error", { "max": 2 }]*/`

pedrottimark commented:
Here are a few potential distractions (to avoid) that went through my mind when I read the examples:
* Declare the variables bar, baz, qux at the beginning of the code (or of a function) and change from variable declarations to assignment statements in blocks of if, else, for.
* To make example code parallel, provide exactly one of each: function declaration, variable initialization to function expression, immediately-invoked function expression.
* In the if-else and switch-case, how about bar = 1 and bar = 2 instead of bar = 1 in both code paths?

scriptdaemon commented:
@pedrottimark Thank you or your feedback.

I've added the changes based on your feedback and removed some unnecessary examples to keep the doc clear and concise. Let me know if I have forgotten something or need to make further changes.

The docs were a bit inconsistent when writing up this rule, so I was a little confused as to which to base mine.

pedrottimark commented:
The changes based on feedback LGTM

@scriptdaemon Ha, only a bit inconsistent? You did well. I look forward to using this rule. And I forgot to say, welcome to ESLint!

nzakas commented:
Thanks @scriptdaemon!
